internal/controller/dnsrecord_controller.go

```diff
@@ -164,34 +173,83 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		return ctrl.Result{RequeueAfter: randomizedValidationRequeue}, nil
 	}
 
-	if !common.Owns(managedZone, dnsRecord) {
-		logger.V(1).Info("Record is not owned by ManagedZone", "ManagedZoneName", managedZone.Name)
-		err = common.EnsureOwnerRef(managedZone, dnsRecord, true)
+	err = dnsRecord.Validate()
+	if err != nil {
+		logger.Error(err, "Failed to validate record")
+		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse,
+			"ValidationError", fmt.Sprintf("validation of DNSRecord failed: %v", err))
+		return r.updateStatus(ctx, previous, dnsRecord, false, err)
+	}
+
+	// Ensure ZoneID and ZoneDomainName are set in the status
+	// Handles both the case where we have never assigned a zone to a DNSRecord (first time reconcile) and when the
+	// `rootHost` has changed and potentially needs assigned to a different zone.
+	if dnsRecord.Status.ZoneID == "" || dnsRecord.Spec.RootHost != dnsRecord.Status.RootHost {
+		logger.Info(fmt.Sprintf("root host changing from %s to %s, updating zone config",
+			dnsRecord.Status.RootHost, dnsRecord.Spec.RootHost))
+
+		// Create a dns provider with no config to list all potential zones available from the configured provider
+		dnsProvider, err := r.ProviderFactory.ProviderFor(ctx, dnsRecord, provider.Config{})
 		if err != nil {
-			return ctrl.Result{}, err
+			setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse,
+				"DNSProviderError", fmt.Sprintf("The dns provider could not be loaded: %v", err))
+			return r.updateStatus(ctx, previous, dnsRecord, false, err)
 		}
-		err = r.Client.Update(ctx, dnsRecord)
-		return ctrl.Result{}, err
+
+		z, err := dnsProvider.DNSZoneForHost(ctx, dnsRecord.Spec.RootHost)
```

Comment:
Should we do this or should we just expect the zoneid to be specified? Wondering a user wouldn't just specify the zone id and instead allow us to find it?

Reply:
Based on conversation around this, decided to go with removing the secret filter options and instead relying on the credential to be limited down to what we should see. We will add docs giving examples of how to do this. This will mean we can avoid having things sepecified in two places (in the secret config and in the actual IAM policies etc). If we find two zones for the same domain, we will error out as we don't want to just choose one. In this case the permissions the credential has should be modified to remove the duplicate